Scripts/Naxxramas: Rework of the Gluth encounter

chaodhib opened the pull request:
Fixes of the follow bugs:

1) Gluth should eat the zombies close to him (https://github.com/TrinityCore/TrinityCore/issues/14806). The animation used now is not 100% blizzlike but it's an improvement nonetheless.
2) The zombies should put the "Infected Wound" debuff on their auto-attack targets.
3) Zombies should have a normal threat table before decimate and should loose all kind of aggro behavior and run toward Gluth after decimate.
4) Gluth should be not be affected by the decimate damage.
5) Newly poped zombies should have all players in their aggro list as soon as they spawn. So far, they were not moving and were only aggroed by proximity.
6) and several minor fixes (emotes, useless data in db, ...).

DDuarte commented:
Don't hardcode the database name in the SQL file. We do not force any specific DB name.
Remove '' from numbers in the SQL.

DDuarte commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -3660,17 +3660,7 @@ void Spell::EffectScriptEffect(SpellEffIndex effIndex)
-                    }
-                    return;
-                // Mirren's Drinking Hat
+				// Mirren's Drinking Hat
```
tabs

DDuarte commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -109,48 +106,167 @@ class boss_gluth : public CreatureScript
-                        for (int32 i = 0; i < RAID_MODE(1, 2); ++i)
-                            DoSummon(NPC_ZOMBIE, PosSummon[rand32() % RAID_MODE(1, 3)]);
-                        events.ScheduleEvent(EVENT_SUMMON, 10000);
+                case EVENT_WOUND:
```
The indentation was correct, don't change it.

chaodhib replied:
this was done by VS's auto format but ok.

Naios replied:
msvc is wrong when formatting switches, just insert another 4 spaces and it should be fine.

chaodhib replied:
sure.

DDuarte commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -109,48 +106,167 @@ class boss_gluth : public CreatureScript
+        }
+
+    private:
+        Creature* zombie;
```
Never store Unit* in scripts, use guids instead.

chaodhib replied:
Alright. May I ask why though?

DDuarte replied:
It has caused many issues in the past due to dangling references, thread races and so on.
Also, it seems that most of your uses of `Creature* zombie` can and should be replaced by local variables.

DDuarte commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -35,21 +38,28 @@ enum Creatures
-#define EMOTE_NEARBY    " spots a nearby zombie to devour!"
+#define EMOTE_NEARBY    "Gluth spots a nearby zombie to devour!"
+#define EMOTE_DECIMATE  "Gluth decimates all nearby flesh!"
+#define EMOTE_ENRAGE    "Gluth becomes enraged!"
```
Since the code is being changed it might be a good idea to move these texts to the DB (Talk() and `creature_text`)

Nyeriah commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -109,48 +106,167 @@ class boss_gluth : public CreatureScript
+        {
+            if (GetOriginalCaster()->GetEntry() == NPC_GLUTH)
+            {
+                boss_gluth::boss_gluthAI* bossAI = (boss_gluth::boss_gluthAI*)GetOriginalCaster()->ToCreature()->GetAI();
```
Use DoAction() instead of AI casts

Nyeriah commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -35,21 +38,28 @@ enum Creatures
+#define EMOTE_ENRAGE    "Gluth becomes enraged!"
+
+#define SECONDS *1000
+#define MINUTES *60000
```
There are already definitions for this, in general it goes by XXXX * MINUTES * IN_MILLISECONDS

example usage: https://github.com/TrinityCore/TrinityCore/blob/d7c53c30fb45f6da015e126aeb8a4a1901f72e58/src/server/scripts/Outland/TempestKeep/Mechanar/boss_mechano_lord_capacitus.cpp#L85

chaodhib replied:
To be honest,
```C++
events.ScheduleEvent(EVENT_SUMMON, 15 SECONDS);
events.ScheduleEvent(EVENT_BERSERK, 8 MINUTES);
```
 is far more clear to read and write than 
```C++
events.ScheduleEvent(EVENT_SUMMON, 15 * IN_MILLISECONDS);
events.ScheduleEvent(EVENT_BERSERK, 8  * MINUTE * IN_MILLISECONDS);
```
respectively.  It also add this abstraction where the scripter doesn't have to remember that the second parameter of ScheduleEvent() is in milliseconds. And stuff like this is also possible AND clean at the same time:

```C++
#define MILLISECONDS *1
#define SECONDS *1000
#define MINUTES *60000
#define HOURS *60*60000

events.ScheduleEvent(EVENT_BERSERK, 2 MINUTES + 25 SECONDS);
```

But I get that it may create name conflicts and also that making the change would mean changing every script which I'm not volunteering to do at the moment :p So, I'll make the change if needed.

DDuarte replied:
Once we support C++14 we can use duration literals (http://en.cppreference.com/w/cpp/chrono/duration#Literals). Until then, don't try to reinvent them.

Shauren replied:
No, stick to what we currently have - also noone has to remember that it is in milliseconds, in fact it is well documented on ScheduleEvent definition

Naios replied:
User defined literals are already supported in c++11, ( this will support literals like `13_s` or `2_min` becaus ethe underscore is forced) - only supported in gcc, clang and msvc 2015.
C++14 introduces duration literals which don't need an underscore anymore `13s` or `2min`.

But for now you can try to use the [TaskScheduler](https://gist.github.com/Naios/6d143538abdc3d32fd2b) which ships with some nice time typedefs like `Seconds(13), Minutes(2)` so you can drop your custom defines:

```c++
scheduler.Schedule(Minutes(2) + Seconds(25), [this](TaskContext context)
{
    // do something
});
```

Nyeriah commented on sql/updates/world/2015_06_08_10_world.sql:
```diff
@@ -0,0 +1,14 @@
+-- the spell1 column was missused for this NPC. This NPC is undead. He cannot be MCed. And the spell 29307 should be casted using SmartAI.
+UPDATE `world`.`creature_template` SET `spell1`='0' WHERE `entry`='16360';
```
Don't define database names in SQL files

Kittnz commented:
Nice work none the less!

tkrokli commented:
Just in case you want a quick fix for your SQL, here is a modified and updated version:
```sql
-- The column `spell1` was misused for this NPC. This NPC is undead. He cannot be MCed.
-- And the spell 29307 should be cast using SmartAI.
UPDATE `creature_template` SET `spell1`=0 WHERE `entry`=16360;
UPDATE `creature_template` SET `spell1`=0 WHERE `entry`=30303;

-- The zombie chows cast the spell 29307 on themselves which makes their AA apply one stack of the debuff "Infected Wound".
DELETE FROM `smart_scripts` WHERE `entryorguid`=16360;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(16360, 0, 0, 0, 4, 0, 100, 31, 0, 0, 0, 0, 11, 29307, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Zombie Chow - On Creature Aggro - Cast \' Infected Wound\' on self (No Repeat)');
UPDATE `creature_template` SET `AIName`='SmartAI' WHERE `entry`=16360;

-- connect the decimate spell (effect 0 & 2) used by Gluth (in both 10 man & 25) to the spell script defined in C++
DELETE FROM `spell_script_names` WHERE `spell_id` IN (28374, 54426);
INSERT INTO `spell_script_names` (`spell_id`, `ScriptName`) VALUES
(28374, 'spell_gluth_decimate'),
(54426, 'spell_gluth_decimate');

chaodhib commented:
@tkrokli thanks! I will use this.

joschiwald commented on sql/updates/world/2015_06_08_10_world.sql:
```diff
@@ -0,0 +1,14 @@
+-- The zombie chows cast on themselves the spell 29307 which makes their AA apply one stack of the debuff "Infected Wound".
+DELETE FROM `world`.`smart_scripts` WHERE `entryorguid`='16360';
+INSERT INTO `world`.`smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `targe ...
+UPDATE `world`.`creature_template` SET `AIName`='SmartAI' WHERE `entry`='16360';
```
you can do this with `creature_template_addon`?

tkrokli replied:
No, the `creature_template_addon` (as well as `creature_addon`) table does not support `smart_scripts`.
That table only supports: `entry`, `path_id`, `mount`, `bytes1`, `bytes2`, `emote` and `auras` .

DDuarte replied:
@tkrokli: What joshchiwald means is that this spell cast can and should be moved to `creature_template_addon.auras` since it is an aura cast on reset/spawn.

tkrokli replied:
Sorry, I didn't see the connection. I will delete my comments.

Killyana replied:
This is wrong, every time you get hit by a zombie you will get 1 Infected Wound (99 is the max)
In this case you must use: SMART_EVENT_DAMAGED_TARGET with the target 2 and not 1

chaodhib replied:
The aura cast on spawn has been moved in script now.

chaodhib commented:
I think the script is pretty complete now. Every mechanics of the boss are in there. But it may need debugging. 

So, in need of feedback please. Both in code review and actual testing. Thanks

tkrokli commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -1,55 +1,83 @@
- *
- * You should have received a copy of the GNU General Public License along
- * with this program. If not, see <http://www.gnu.org/licenses/>.
- */
```
Like Aokromes said, keep the top section unchanged and with the original spacing.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+            if (!UpdateVictim())
+                return;
+
+            Creature* gluth = ObjectAccessor::GetCreature(*me, GluthGUID);
```
this is not necessary

chaodhib replied:
line 391. there is a need for a reference to gluth's Creature*. I guess I could move that line into the if at 387 though.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+                        me->GetMotionMaster()->MovePoint(0, gluth->GetPosition()); // isnt dynamic. point to point move without taking into account that Gluth might change his position before reaching destination point.
+                        //me->GetMotionMaster()->MoveFollow(gluth, 0, 0); // the zombies will stand still too far away from Gluth at the end of their movement. Also doesnt seem to work with SetWalk(true)
+                        //me->GetMotionMaster()->MoveChase(gluth); // crash server. I assume its because Gluth is friendly. We need the zombies to get really close to Gluth's center position anyway (10 yards) for the multi kill spell to work.
+
```
try to use an event every 1 or 2 seconds with 

````c++
me->GetMotionMaster()->MovePoint(0, gluth->GetPosition());
````

that'll do the trick i think

chaodhib replied:
sure, I'll test.

chaodhib replied:
I love you. It works pretty nicely.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+
+        void SetData(uint32 id, uint32 value) override
+        {
+            if (id == 1) // change of state
```
as some colabs would say.... dont use magic numbers LOL

chaodhib replied:
i'm gonna remove the second parameter and rename the first parameter as "value" since there is only one use of SetData. Should be clearner. I want to avoid creating an Enum for just on element.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+        {
+            if (index == 1)
+                return state;
+            return -1;
```
return type is unsigned, use 0.

chaodhib replied:
ok

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+        
+        bool Validate(SpellInfo const* /*spellInfo*/) override
+        {
+            if (sSpellMgr->GetSpellInfo(SPELL_ZOMBIE_CHOW_SEARCH_MULTI) && sSpellMgr->GetSpellInfo(SPELL_ZOMBIE_CHOW_SEARCH_SINGLE))
```
it should be "||" instead of "&&", shouldn't it?

chaodhib replied:
Valid should be removed actually. Both spell will call this spell script. If this spell script is called, this automaticly mean the spell that called it exists in the db. No need to check for the existence of another spell. I misused Validate() on that one.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+        // This method handles the change of the zombie's behavior after a decimate.
+        void HandleSendEvent(SpellEffIndex /* index */)
+        {
+            if (GetOriginalCaster()->GetEntry() == NPC_GLUTH)
```
originalCaster is not always referenced, use a null pointer check

chaodhib replied:
originalCaster can be null? How?

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+            if (Unit *unit = GetHitUnit())
             {
-                if (me->IsWithinMeleeRange(me->GetVictim()))
+                if (unit->GetEntry() == NPC_ZOMBIE || unit->GetTypeId() == TypeID::TYPEID_PLAYER || unit->IsPet())
```
it's not necessary to use "TypeID::"

chaodhib replied:
ok

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+                    Creature* zombie = nullptr;
+                    for (SummonList::const_iterator itr = summons.begin(); itr != summons.end(); ++itr)
+                    {
+                        zombie = ObjectAccessor::GetCreature(*me, *itr);
```
use an if block to avoid a null pointer reference

chaodhib replied:
this comes for the summon list. there is no null reference in the summon list. If a zombie dies, line 121 is used to remove it from the list. if the list is empty, itr == summons.end() equals true as soon as the loop starts therefore no problem of null pointer.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+
+                        // and then return on primary target
+                        me->SetReactState(ReactStates::REACT_AGGRESSIVE);
+                        me->SetFacingToObject(nullptr);
```
set facing to target?

chaodhib replied:
This line is a bit akward. I'm not really sure how it works. I wanted to null the effect of line 201 since I noticed when Gluth is stunned (by the effect of Decimate), he would turn toward the zombie he previously killed. But it was a bug in an older version. I'm not sure it's still there. And I'm not sure if this line do anything anyway.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+
+                        Creature* zombie = nullptr;
+                        for (SummonList::const_iterator itr = summons.begin(); !zombie && itr != summons.end(); ++itr)
+                            if (ObjectAccessor::GetCreature(*me, *itr)->GetExactDist2d(me) <= 18.0)
```
null pointer check >.<

chaodhib replied:
SummonList shouldn't have anything null. See other comment.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+                        Creature* zombie = nullptr;
+                        for (SummonList::const_iterator itr = summons.begin(); !zombie && itr != summons.end(); ++itr)
+                            if (ObjectAccessor::GetCreature(*me, *itr)->GetExactDist2d(me) <= 18.0)
+                                zombie = ObjectAccessor::GetCreature(*me, *itr);
```
i think u should improve this search.

here is an example:

there are 4 zombies, only one if them is in range:
the summonlist could be like this:

zombie1 (not in range), zombie 2 (in range), zombie3 (not in range), zombie 4 (not in range)

like this, for will always iterate along the list till last position. So the value of "Creature* zombie = nullptr;" will only depend on last position, which is not a correct behaviour.

I suggest a while block + a boolean to stop it, in case a valid case is found.

ccrs commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+                    {
+                        Creature* zombie = nullptr;
+                        for (SummonList::const_iterator itr = summons.begin(); !zombie && itr != summons.end(); ++itr)
+                            if (ObjectAccessor::GetCreature(*me, *itr)->IsWithinDistInMap(me, 10.0))
```
null pointer check again

chaodhib replied:
SummonList shouldn't have anything null. See other comment.

Nyeriah replied:
Some strong hunch tells me the spell 28239 should handle the target searching.

Treeston replied:
Cache ObjectAccessor::GetCreature returns, please.
Also, there is no core guarantee that guids are still valid at any given time, so nullptr check this.

ccrs commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -3660,16 +3660,6 @@ void Spell::EffectScriptEffect(SpellEffIndex effIndex)
                 }
-                // Decimate
-                case 28374:
-                case 54426:
```
This other id is from ICC, is it ok to remove it?

Shauren replied:
No, it is no, ICC uses 71123

chaodhib replied:
and this code shouldnt be there anyway, right? Spreaded code like this is a nightmare to maintain.

chaodhib commented on sql/updates/world/2015_06_08_10_world.sql:
```diff
@@ -0,0 +1,27 @@
+
+-- connect the decimate spell (effect 0 & 2) used by Gluth (in both 10 man & 25) to the spell script defined in boss_gluth.cpp.
+-- and connect the 2 zombie chow search spells to a unique spell scripted use by both.
+DELETE FROM `spell_script_names` WHERE `spell_id` IN (28374, 54426);
```
add 28239 & 28404 in the IN clause.

Shauren replied:
no, **always** delete by `ScriptName` not by `spell_id` because a single spell may have more than one script (and you are adding a script)

chaodhib replied:
True. However, now that I think about it, the opposite is also true: A script can be called by more than one spell (I used one of those here). Therefore, I suppose I should specify both `ScriptName` & `spell_id`?

chaodhib commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +133,333 @@ class boss_gluth : public CreatureScript
+    }
+
+};
+
```
add comment:
// called by both 28239 & 28404 to heal gluth for each target hit

Nyeriah replied:
Why both? 28404 is the one with the instakill effect.

chaodhib commented on sql/updates/world/2015_06_08_10_world.sql:
```diff
@@ -0,0 +1,27 @@
+(28404, 'spell_gluth_zombiechow_search');
+
+-- add the condition to the multi-target insta-kill spell 28404 which is that only zombies are eligible targets to the spell.
+DELETE FROM `conditions` WHERE `SourceEntry`=28404;
```
Any way to make this query safe or does it not matter?

Shauren replied:
Add `SourceTypeOrReferenceId`=13 to WHERE clause

chaodhib replied:
ok thanks

Shauren commented on sql/updates/world/2015_06_08_10_world.sql:
```diff
@@ -0,0 +1,27 @@
+-- The column `spell1` was misused for this NPC. This NPC is undead. He cannot be MCed. The spell 29307 should be cast using SmartAI/ by Script.
+-- Also, link the zombies chows to the custom ScriptedAI defined in boss_gluth.cpp.
+UPDATE `creature_template` SET `spell1`=0, `ScriptName`='npc_zombie_chow' WHERE `entry`=16360;
+UPDATE `creature_template` SET `spell1`=0, `ScriptName`='npc_zombie_chow' WHERE `entry`=30303;
```
difficulty entries should not have a scriptname set

chaodhib replied:
ok

chaodhib commented:
I noticed Decimate has an EffectMiscValueA = 10495 on its third effect SPELL_EFFECT_SEND_EVENT. I would like to use the DB to handle this effect instead of using another hook on the spell script (that calls DoAction() on Gluth which calls SetData() on each zombie). 

I think I need to use the event_scripts table. I would like the event to SetData(1, 3) on each zombie. It seems like I could use the command SCRIPT_COMMAND_FIELD_SET but I dont know how/where to define the creatures that will be affected by the event. 

Any ideas?

chaodhib commented:
I'm gonna try to rework the Decimate spell (28374). I noticed recently that:
* Effect 0 would be best suited to handle the change in the zombies' behavior since it has Targets set to TARGET_SRC_CASTER, TARGET_UNIT_SRC_AREA_ENTRY. A row in conditions table specifying the zombie' entry and a very short spellscript onHit hook with SetData(1,STATE_ZOMBIE_DECIMATED) should do the trick (or the equivalent in the spell_scripts table if I can manage to do it).
* so Effect 2 should handle the damage component (it should put everyone except Gluth to 5% hp). Also, EffectMiscValueA = 10495 hasnt be used so far. I would assume this should reference a row in the event_scripts table.

buuuut, this process of moving script code to the DB is time consuming. In the end, my question is: Is it even worth it to put as much behavior and information as possible in the DB? Is it something considered good practise?

chaodhib commented:
Thanks for the help & feedback on IRC. Any last changes before I squash the commits?

Nyeriah commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -63,39 +81,36 @@ class boss_gluth : public CreatureScript
         {
-            // Do not let Gluth be affected by zombies' debuff
-            me->ApplySpellImmune(0, IMMUNITY_ID, SPELL_INFECTED_WOUND, true);
+            zombieToBeEatenGUID = ObjectGuid::Empty;
```
ObjectGuid does not need to be initialized.

Nyeriah commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -63,39 +81,36 @@ class boss_gluth : public CreatureScript
-            else
-                BossAI::MoveInLineOfSight(who);
+            _Reset();
+            zombieToBeEatenGUID = ObjectGuid::Empty;
```
If you need to reset it ty .Clear()

chaodhib commented:
Bump.

Is there anything missing? bugged? something you suspect is unblizz-like and I should defend?

Would you like to see the notes I took for this boss? Since I couldn't find much information in the sniff given to me, I used a lot of 3.X.Y videos of the encounter, I took note of the spells timers and the positioning of Gluth and the zombies.

mik1893 commented:
i'll test it tomorrow... to me it  looks fine, might be @joschiwald looks into it as he's the script master :D

tkrokli commented on sql/updates/world/2015_06_XX_YY_world.sql:
```diff
@@ -0,0 +1,35 @@
+-- add a condition to the decimate spell (28374/54426). It's damage component should only affect players and zombies.
+DELETE FROM `conditions` WHERE `SourceEntry`=28404 AND `SourceTypeOrReferenceId`=13;
+DELETE FROM `conditions` WHERE `SourceEntry`=28374 AND `SourceTypeOrReferenceId`=13;
+DELETE FROM `conditions` WHERE `SourceEntry`=54426 AND `SourceTypeOrReferenceId`=13;
```
These 3 lines can be merged into one like this:
```sql
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`= 13 AND `SourceEntry` IN (28404, 28374, 54426);
```
This can be done when only 1 column contains multiple variables. Not a big deal, since your version does the same.
It is practical for the sake of overview to refer the column names in the order they appear in the DB table, though.

chaodhib replied:
Ok thanks.

tkrokli commented:
Ouch. I just noticed that `chaodhib:3.3.5` is 163 commits behind `TrinityCore:3.3.5` (25 days since last update). I think this could be the reason for some or all of the conflicts. If you didn't know, use the Rebase command or Rebase option when pulling the TC 3.3.5 source again (updating your local copy), to keep your own changes on top of the updated TC source. :+1:

chaodhib commented:
> Remember to add your comment to the file boss_gluth.cpp before your next commit push to this PR.

What do you mean? What comment?

About the conflict, it's caused by a rename made by Shauren https://github.com/TrinityCore/TrinityCore/commit/b4adf3f5dd34d8171f10a6a4ee8689286ef4d66a. It doesn't matter, it's easy to solve.

chaodhib commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +120,309 @@ class boss_gluth : public CreatureScript
+                    me->StopMoving();
+                    me->GetMotionMaster()->Clear();
+                    me->GetMotionMaster()->MoveIdle();
+                    me->ApplySpellImmune(0, IMMUNITY_MECHANIC, MECHANIC_GRIP, true); // not sure if this is blizz-like but this is very convenient
```
Unsure about this. I'm not really sure what happens in retail when a zombie that is about to be eaten gets death griped.

if this line is remove, Gluth will follow the zombie.

tkrokli commented:
This comment:
>  [chaodhib added a note on 19 Jun](https://github.com/TrinityCore/TrinityCore/pull/14865#discussion-diff-32854482) (<= link to the post where you commented on the code)
add comment:
// called by both 28239 & 28404 to heal gluth for each target hit

(I just assumed that you wanted to remind yourself to add it in the code.)

chaodhib commented:
Oh that. It's already done. I don't know why github doesn't consider this as comment on an outdated dif.

tkrokli commented:
Not sure either. Just manually delete the comment in the code to mark that it is done. :-)

chaodhib commented:
Since I made the comment on a blank line and this blank line isnt modified, github doesn't recognize that the comment is no longer applicable I guess. I'm just gonna leave it like that.

DjYXA commented:
Is this ready to pull?

chaodhib commented:
Yes.

I should probably read in detail @joschiwald's Thorim rework and see if I could get ideas to improve my code but this rework is already functional and finished as far as I can tell.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -110,47 +120,309 @@ class boss_gluth : public CreatureScript
+                Creature* gluth = ObjectAccessor::GetCreature(*me, GluthGUID);
+                // Putting this in the UpdateAI loop fixes an issue where death gripping a decimated zombie would make the zombie stand still until the rest of the fight.
+                // Also fix the issue where if one or more zombie is rooted when decimates hits (and MovePoint() is called), the zombie teleport to the boss. pretty weird behavior.
+                if (me->GetExactDist2d(gluth) > 10.0 && me->CanFreeMove() && timer>1600) // it takes about 1600 ms for the animation to cycle. This way, the animation looks relatively smooth.
```
Do the timer check first, that way distance calculation doesn't need to be done most of the time.

chaodhib replied:
Yep good idea

chaodhib commented:
The single kill zombie animation ("Gluth spots a nearby zombie to devour") needs more work. It's functional but kinda ugly. Sniffs for that would be appreciated.

Other than that, I'm not quite sure why this PR isn't getting merged. It's been 2 months now...

StormBytePP commented:
We don't merge things that just works unless they do in a correct way. You
have answered yourself.
El 15/9/2015 4:21, "Chaouki Dhib" <notifications@github.com> escribió:

> The single kill zombie animation ("Gluth spots a nearby zombie to devour")
> needs more work. It's functional but kinda ugly. Sniffs for that would be
> appreciated.
>
> Other than that, I'm not quite sure why this PR isn't getting merged. It's
> been 2 months now...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TrinityCore/TrinityCore/pull/14865#issuecomment-140257849>
> .
>

chaodhib commented:
I don't plan on making any more improvement on this boss (since I don't have the time anymore). Either this get merged and someone else tries to fix the animation in a later commit or this PR should be closed.

If no response in 2 weeks, I'm closing this PR either way.

I would like to insist that this commit makes the boss actually playable and 95% blizzlike. It's a big improvement compared to the current version of this boss. I don't think it's fair to reject this commit because the result is not 100% perfect. Incremental changes is a pretty standard way of building things in software engineering...

Treeston commented:
@chaodhib give me push powers on the branch, please

chaodhib commented:
@Treeston Done

Treeston commented:
Rebased and conflicts resolved. If I get some free time, I'll look into the animation thing.

chaodhib commented:
Your help is gladly welcome. Thanks.

jackpoz commented:
There are new conflicts in this PR

tkrokli commented:
chaodib needs to rebase his 3.3.5 fork to solve it, 58 commits behind the TC 3.3.5 branch.

chaodhib commented:
Done.

Treeston commented:
Rebase done.

tkrokli commented on src/server/scripts/Northrend/Naxxramas/naxxramas.h:
```diff
@@ -189,6 +196,9 @@ enum InstanceEvents
     EVENT_DIALOGUE_SAPPHIRON_LICHKING2,
     EVENT_DIALOGUE_SAPPHIRON_KELTHUZAD3,
     EVENT_DIALOGUE_SAPPHIRON_KELTHUZAD4
```
Missing comma after this line when adding the next 3 lines.

chaodhib commented:
@Treeston I'll fix this. I already have a completed rebase ready.

Treeston commented:
Alright - please try not to duplicate enums too much; having multiple Event ID enums in the .h is redundant, merge them into one enum.

chaodhib commented:
InstanceEventIDs != InstanceEvents. InstanceEventIDs values are from dbc. The distinction is made in other instance headers IIRC.

Treeston commented:
There's actually a dbc for that? TIL

chaodhib commented:
No i mean that value is taken from the spell 28374 in the Spell.dbc file.

jackpoz commented:
@Treeston is this still a work in progress ? how's going with the animation thing ?

if it's just a cosmetic issue that still improves the current sources it could be merged.

Treeston commented:
I'll probably get around to it during the holidays.

ShinDarth commented:
are there any news about this?

ccrs commented:
it needs to be updated I guess

Treeston commented:
Yeah, it does. I was busy with other PRs and didn't get around to this one yet. No promises as to when I'll get to.

Treeston commented on src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:
```diff
@@ -182,6 +182,12 @@ class instance_naxxramas : public InstanceMapScript
+                        GluthGUID = creature->GetGUID();
+                        break;
+                    case NPC_ZOMBIE_CHOW:
+                        ZombieChowGUIDs.push_back(creature->GetGUID());
```
Why do you rely on InstanceScript in order to do something that should be handled by the BossAI. Do not clutter instance events like this, please.

Treeston commented on src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:
```diff
@@ -218,6 +224,23 @@ class instance_naxxramas : public InstanceMapScript
                         else
                             events.ScheduleEvent(EVENT_THADDIUS_RESET, 30 * IN_MILLISECONDS);
                         break;
+                    case EVENT_GLUTH_ZOMBIE_BEHAVIOR: // handles the change in the zombies' behavior after a decimate.
```
Get a handle to them using SummonList, then handle behavior change in their AI. Don't do this in the instance script.

Treeston commented on src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:
```diff
@@ -315,13 +338,20 @@ class instance_naxxramas : public InstanceMapScript
+                                kelthuzad->AI()->Talk(SAY_KELTHUZAD_CAT_DIED);
+                            break;
+                        case NPC_ZOMBIE_CHOW:
+                            ZombieChowGUIDs.remove(creature->GetGUID());
```
As above. This should be handled in Gluth's AI.

Treeston commented on src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:
```diff
@@ -344,6 +374,10 @@ class instance_naxxramas : public InstanceMapScript
                     case DATA_HAD_FAERLINA_GREET:
                         hadFaerlinaGreet = (value == 1u);
                         break;
+                    case DATA_GLUTH_ZOMBIES_RESET:
```
As above.

Treeston commented on src/server/scripts/Northrend/Naxxramas/naxxramas.h:
```diff
@@ -43,6 +43,10 @@ enum Encounter
 {
+    // Gluth
+    DATA_GLUTH_ZOMBIES_RESET,
+    DATA_ZOMBIE_STATE,
```
As above.

Treeston commented on src/server/scripts/Northrend/Naxxramas/naxxramas.h:
```diff
@@ -189,7 +196,10 @@ enum InstanceEvents
+    EVENT_DIALOGUE_SAPPHIRON_KELTHUZAD4,
+    
+    // Gluth
+    EVENT_GLUTH_ZOMBIE_BEHAVIOR = 10495 // event that will trigger the zombies' behavior change after Decimate.
```
As above.

Treeston commented on src/server/scripts/Northrend/Naxxramas/naxxramas.h:
```diff
@@ -212,6 +222,14 @@ enum InstanceTexts
 };
 
+enum InstanceStates
+{
```
Should be in boss_gluth.

chaodhib commented:
@Treeston 
I'm only using the instance script in order to have access to the `ProcessEvent `hook for the decimate spell (28374). I used to have everything regarding that spell in boss_gluth.cpp but I think that the code is a bit cleaner now:

Gluth cast the spell and it leaves its hand. The rest is handle elsewhere then `boss_gluthAI`. No need to make `boss_gluthAI `call the spell which calls a spell script which calls yet another method of `boss_gluthAI `which calls a method on `npc_zombie_chowAI`.

On the other hand, I completely agree with you that this solution sucks too. This stuff should not pollute instance_naxxramas.cpp when its only linked to Gluth.

In definitive, I think it would be pretty cool to have access to a ProcessEvent hook from the boss AIs. Is it possible?

Treeston commented:
Why don't you just use `SpellHit` on the zombies? No need to really use `ProcessEvent`. Yes, even if Blizzard does it. Blizzard does weird things at times.

chaodhib commented:
> Blizzard does weird things at times

This is usually because they are not well understood (and believe me, I know it's not easy). My theory is that the SEND_EVENT spell effect is handled a bit differently in retail servers, which makes its use easy and clean. Maybe each boss script has its own `ProcessEvent` hook which is registered into the instance where the boss is.

Anyway, for this particular spell, you are right, a `SpellHit ` hook will do it the trick so I'll do that, until (if) `ProcessEvent `hooks are reworked. Thanks

chaodhib commented:
Done. Anything else?

Treeston commented on src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:
```diff
@@ -481,7 +488,6 @@ class instance_naxxramas : public InstanceMapScript
 
                 return true;
             }
-
```
Leave the blank line between methods.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                else if (value == STATE_ZOMBIE_TOBE_EATEN)
+                {
+                    // loose aggro behavior
+                    me->SetReactState(ReactStates::REACT_PASSIVE);
```
No need to specify `ReactStates::`

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                    // and forced to stand still
+                    me->StopMoving();
+                    me->GetMotionMaster()->Clear();
+                    me->GetMotionMaster()->MoveIdle();
```
No need to specifically `MoveIdle` here.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
                 {
-                    me->Kill(me->GetVictim());
-                    me->ModifyHealth(int32(me->CountPctFromMaxHealth(5)));
+                    me->GetMotionMaster()->MovePoint(0, gluth->GetPosition()); // isn't dynamic. So, to take into account Gluth's movement, it must be called periodicly.
```
Why not `MoveChase`?

chaodhib replied:
It's not working. I'm not sure why. Possibly because Gluth and the zombies are seen as "friendly" to each other faction wise.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                timer += diff;
+                Creature* gluth = ObjectAccessor::GetCreature(*me, GluthGUID);
+                // Putting this in the UpdateAI loop fixes an issue where death gripping a decimated zombie would make the zombie stand still until the rest of the fight.
+                // Also fix the issue where if one or more zombie is rooted when decimates hits (and MovePoint() is called), the zombie teleport to the boss. pretty weird behavior.
```
it's a movespline bug, see #16681

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
 
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return new npc_zombie_chowAI(creature);
```
Use [`GetInstanceAI`](https://github.com/Treeston/TrinityCore/blob/3.3.5-naxxcleanup/src/server/scripts/Northrend/Naxxramas/boss_thaddius.cpp#L167) to make sure `me->GetInstanceScript()` is always not null. Otherwise, core crash if you add the NPC outside of an instance.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -63,39 +93,33 @@ class boss_gluth : public CreatureScript
-                BossAI::MoveInLineOfSight(who);
+            _Reset();
+            zombieToBeEatenGUID.Clear();
+            state = STATE_GLUTH_NORMAL;
```
Reset react state.

chaodhib replied:
what do you mean? it's done one line below.

Treeston commented on sql/updates/world/AAAA_ZZ_XX_YY_world.sql:
```diff
@@ -0,0 +1,32 @@
+-- adds Gluth's 5 emotes into the DB.
+DELETE FROM `creature_text` WHERE `entry`=15932 AND `groupid` IN(0,1,2,3,4) AND `id`=0;
+INSERT INTO `creature_text` (`entry`, `groupid`, `id`, `text`, `type`, `language`, `probability`, `emote`, `duration`, `sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
+(15932, 0, 0, '%s spots a zombie to devour!', 16, 0, 100, 0, 0, 0, 12242, 1, 'Gluth - spots one'),
```
Textrange should (almost) always be 2/3 for boss emotes.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -18,38 +18,68 @@
+    // Gluth
+    SPELL_MORTAL_WOUND                  = 54378, // spell effect dummy unused. what its supposed to do is unkown.
+    SPELL_ENRAGE                        = 28371, // 54427 in 25 man.
+    SPELL_DECIMATE                      = 28374, // 54426 in 25 man. Effect0 is handled by SpellScript (see below) and 2 rows in conditions table. Effect2 is handled by InstanceScript (see instance_naxxramas.cpp).
```
No longer handled by instance script, update comment.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -18,38 +18,68 @@
+    SPELL_ZOMBIE_CHOW_SEARCH_MULTI      = 28404, // Insta kill spell. Affect all zombies 10 yards around Gluth's center. See one row conditions table + spell script below.
+
+    // Zombies
+    SPELL_INFECTED_WOUND                = 29307  // Used by the zombies on self.
```
Move this to `creature_template_addon`

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -63,39 +93,33 @@ class boss_gluth : public CreatureScript
         {
-            // Do not let Gluth be affected by zombies' debuff
-            me->ApplySpellImmune(0, IMMUNITY_ID, SPELL_INFECTED_WOUND, true);
+            state = STATE_GLUTH_NORMAL;
```
Move to initializer list.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                        state = STATE_GLUTH_NORMAL;
+
+                        // and then return on primary target
+                        me->SetReactState(ReactStates::REACT_AGGRESSIVE);
```
No need to specify REACT_AGGRESSIVE.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+
+                        // and then return on primary target
+                        me->SetReactState(ReactStates::REACT_AGGRESSIVE);
+                        me->SetTarget(me->GetVictim()->GetGUID());
```
Use `MustReacquireTarget()` to force him to reselect target in `Attack`.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                        // and then return on primary target
+                        me->SetReactState(ReactStates::REACT_AGGRESSIVE);
+                        me->SetTarget(me->GetVictim()->GetGUID());
+                        me->GetMotionMaster()->MoveChase(me->GetVictim());
```
Not 100% on this, but I think attack code should do this automatically.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                        Creature* zombie = nullptr;
+                        for (SummonList::const_iterator itr = summons.begin(); !zombie && itr != summons.end(); ++itr)
+                        {
+                            zombie = ObjectAccessor::GetCreature(*me, *itr);
```
nullptr check this.

ccrs replied:
no need to

Treeston replied:
Why?

chaodhib replied:
done

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+                Creature* gluth = ObjectAccessor::GetCreature(*me, GluthGUID);
+                // Putting this in the UpdateAI loop fixes an issue where death gripping a decimated zombie would make the zombie stand still until the rest of the fight.
+                // Also fix the issue where if one or more zombie is rooted when decimates hits (and MovePoint() is called), the zombie teleport to the boss. pretty weird behavior.
+                if (timer>1600 && me->GetExactDist2d(gluth) > 10.0 && me->CanFreeMove()) // it takes about 1600 ms for the animation to cycle. This way, the animation looks relatively smooth.
```
`timer > 1600`
feels like hack.

chaodhib replied:
it feels like a hack because it is one. I'm using MovePoint (and it needs to be recalled periodicly) because MoveFollow doesn't give me full control of the exact distance from which the unit stand at.

Treeston commented on src/server/scripts/Northrend/Naxxramas/boss_gluth.cpp:
```diff
@@ -105,52 +129,340 @@ class boss_gluth : public CreatureScript
+            if (id == DATA_ZOMBIE_STATE) // change of state
+            {
+                state = value;
+                if (value == STATE_ZOMBIE_DECIMATED)
```
`switch`?

chaodhib replied:
not really needed here imo

chaodhib committed:
Scripts/Naxxramas: Rework of Gluth

1) Gluth should be able to eat zombies close to him, one at the time, during the whole fight.
2) Gluth should be able to eat zombies at an accelerated rate after each Decimate.
3) The zombies should put the "Infected Wound" debuff on their auto-attack targets.
4) Zombies should have a normal threat table before decimate and should loose all kind of aggro behavior and walk toward Gluth after decimate.
5) Gluth shouldn't be affected by the decimate damage.
6) Newly poped zombies should have all players in their aggro list as soon as they spawn. So far, they were not moving and were only aggroed by proximity.
7) and several minor fixes (emotes, useless data in db).

chaodhib commented:
Done and done. I also fixed the single zombie kill animation. Can this get merged please?

Treeston commented:
For obvious reasons, I'll wait for discussion on #16806 to finish before merging this ;)

msoky commented:
@Treeston https://github.com/TrinityCore/TrinityCore/pull/16806 merged

Treeston commented:
@chaodhib can you FP this so we get travis feedback?

chaodhib commented:
Thanks @Treeston for the std::chrono migration. Should I squash or is there anything to do?

Keader commented:
chrono is magical  :+1:

Treeston commented:
Whoever merges can squash. I'll probably do final review soon.
